plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
 	FuGenesysFwCodesign codesign;
 	GByteArray *st_codesign; /* codesign info, may need to backup for GL352350 */
 	GByteArray *st_public_key;
+
+	/* hid channel */
+	FuGenesysHubhidDevice *hid_channel;
 };
 
 G_DEFINE_TYPE(FuGenesysUsbhubDevice, fu_genesys_usbhub_device, FU_TYPE_USB_DEVICE)
 
+void
+fu_genesys_usbhub_device_set_hid_channel(FuDevice *device, FuDevice *channel)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	if (self == NULL)
+		return;
+
+	if (!FU_IS_GENESYS_HUBHID_DEVICE(channel))
+		return;
+
+	if (self->hid_channel != NULL) {
+		g_warning("already set hid_channel, physical_id(%s)",
+			  fu_device_get_physical_id(FU_DEVICE(self->hid_channel)));
+	} else {
+		self->hid_channel = FU_GENESYS_HUBHID_DEVICE(channel);
+		/* align usb max length(0xffff) to usb2 packet size(0x40) */
+		self->flash_rw_size = 0xffc0;
+	}
+}
+
+static gboolean
+fu_genesys_usbhub_device_send_control_transfer(FuGenesysUsbhubDevice *self,
+					       FuGenesysUsbSetup *setup,
+					       guint8 *data,
+					       FuProgress *progress,
+					       GError **error)
+{
+	if (self->hid_channel != NULL) {
+		guint8 dir = setup->req_type.dir;
+		setup->req_type.dir = !dir; /* revert g_usb in/out dir to usb spec */
```

Comment:
??

Reply:
The define does not match the common USB BM_REQUEST_TYPE https://github.com/fwupd/fwupd/blob/651b6b1207ae923e2368d28c5cd88e02a4ec501d/plugins/genesys/fu-genesys-common.h#L22-L30

`GUsbDeviceDirection` is the inverse of with [wdfusb](https://github.com/microsoft/Windows-Driver-Frameworks/blob/a94b8c30dad524352fab90872aefc83920b98e56/src/publicinc/wdf/kmdf/1.33/wdfusb.h#L50-L56), that cause `G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE` set to 0xC0(vendor read).

Reply:
`FuGenesysUsbRequestType` is using to align the USB spec, that is needed to fulfill format of GL HID feature report. 

The GL HID feature report format has these phases:
1. send request command
`setup`: Set feature report
`data`:  original USB vendor command that we send by g_usb, i.g. `40 81 0100 0000 0000`
2. read/write data
    - `read data`
        - `setup`: Get feature report
        - `data`    
    - `write data`
        - `setup`: Set feature report
        - `data`
3. send end to finish write
`setup`: Get feature report